test/prod_tests.jl

```diff
@@ -327,5 +324,34 @@ end
     @test !insupport(prod, -1)
     @test logpdf(prod, 1) === 2
     @test logpdf(prod, 2) === 4
+end
+
+@testitem "There should be no ambiguity in `prod` for `ProductOf`" begin
+    struct SomeArbitraryDistribution end
+
+    @test prod(
+        GenericProd(),
+        SomeArbitraryDistribution(),
+        ProductOf(SomeArbitraryDistribution(), SomeArbitraryDistribution()),
```

Comment:
If I am not wrong PR tries to fix this as well?

```
julia> prod(GenericProd(), ProductOf(Normal(0, 1), Normal(0, 1)), ProductOf(Normal(0, 1), Normal(0, 1)))
ERROR: MethodError: prod(::GenericProd, ::ProductOf{Normal{Float64}, Normal{Float64}}, ::ProductOf{Normal{Float64}, Normal{Float64}}) is ambiguous.

Candidates:
  prod(::GenericProd, left::ProductOf{L, R}, right::T) where {L, R, T}
    @ BayesBase ~/.julia/packages/BayesBase/MnjV6/src/prod.jl:252
  prod(::GenericProd, left::T, right::ProductOf{L, R}) where {L, R, T}
    @ BayesBase ~/.julia/packages/BayesBase/MnjV6/src/prod.jl:285

Possible fix, define
  prod(::GenericProd, ::T, ::T) where {L, R, L, R, T<:ProductOf{L, R}, T<:ProductOf{L, R}}

Stacktrace:
 [1] top-level scope
   @ REPL[34]:1
  ```

Reply:
Because it still fails on this branch for me.

Reply:
this PR in particular is not aimed to solve this particular case, I think this one can be a separate issue, as it is can be quite complex in general